### PR TITLE
Hot Module Replacement is compatible with Trusted Types.

### PR DESCRIPTION
**Pull request by koto**
**Note:** this is generally just a rebase of https://github.com/webpack/webpack/pull/9861 by @siegrift to reflect changes in React and the Trusted Types API, which has shipped in Chromium 83. This PR only changes the `webpack-4` branch. 

Applications that enforce [Trusted Types](https://web.dev/trusted-types) by using ['require-trusted-types-for'](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/require-trusted-types-for) CSP directive cannot currently use webpack's HMR, as the browser blocks dynamically loading scripts from URLs generated by HMR. That forces the Trusted Types compliant applications (e.g. ones [based on Angular 11+](https://angular.io/guide/security#enforcing-trusted-types), or many [React applications](https://microsoftedge.github.io/edgevr/posts/eliminating-xss-with-trusted-types/)) to either not use webpack HMR at all, or relax the CSP on the dev environment, potentially leading to functional or security bugs only surfacing in production 😨 .

**What kind of change does this PR introduce?**

This PR makes HMR create and use a Trusted Types policy for the code not to violate CSP restrictions, letting the deveopers detect only actual security bugs via CSP violations. 

The policy name can be configured via output options `trustedTypesPolicyName` setting. By default it's `webpack`. 

**Did you add tests for your changes?**

Yes. I've also verified HMR working on the [Angular lazy-loading setup](https://angular.io/guide/lazy-loading-ngmodules) via `yarn link`ing the patched webpack 4.

**Does this PR introduce a breaking change?**

**No**. Trusted Types support is feature-tested. For environments without Trusted Types support (TTs are present in Edge/Chrome/Chromium/Brave, or via a polyfill) the code uses strings as usual - the code is as such backwards-compatible. The only change is a slight increase in HMR loader size; the assumption is that it's only used on dev-environment, so the code changes are not optimized for size.

**What needs to be documented once your changes are merged?**

The new configuration option.

### CODE REVIEW

File: declarations/WebpackOptions.d.ts [hunk 1]
```diff
@@ -1206,6 +1206,10 @@ export interface OutputOptions {
 	 * Handles exceptions in module loading correctly at a performance cost.
 	 */
 	strictModuleExceptionHandling?: boolean;
+	/**
+	 * The name of the trusted types policy created by webpack to serve bundle chunks.
```
Comment (IgorMinar) [inline declarations/WebpackOptions.d.ts:1210]: consider mentioning that the default policy name is `"webpack"`

**Review by IgorMinar — COMMENTED** [review]
This change looks good to me.

It would be awesome to get this reviewed and merged. Trusted Types are already implemented natively in Chrome, Edge, and Opera, as well as in web frameworks like Angular, lit-html, and tools like Karma: https://github.com/w3c/webappsec-trusted-types/wiki/Integrations 

Having webpack onboard would enable many web apps to become much safer and resilient to XSS.

File: declarations/WebpackOptions.d.ts (hunk 1 above)
Comment (koto) [reply declarations/WebpackOptions.d.ts:1210]: Done :)

**Comment by alexander-akait** [comment]
@IgorMinar Can you accept CLA?

**Comment by koto** [comment]
> @IgorMinar Can you accept CLA?

Just did it. Igor is not the author, I am :)

**Comment by sokra** [comment]
This can only be merged after https://github.com/webpack/webpack/pull/9856

Not sure if this should even be merged for webpack 4 since we usually don't want to add features to webapck 4.

**Comment by koto** [comment]
Should I rebase #9856?

**Comment by alexander-akait** [comment]
@koto Yep, will be great

**Comment by Jack-Works** [comment]
#14173 has been merged, what about this PR?

**Comment by alexander-akait** [comment]
This PR for webpack v4

**Comment by vankop** [comment]
@alexander-akait should we close this? I think for webpack@5 it should work already.

**Comment by alexander-akait** [comment]
Yep, I think we don't need it for webpack v4

**Pull request closed without merging**